pkg/util/metric/metric.go

```diff
@@ -1051,3 +1069,223 @@ var RecordHistogramQuantiles = []Quantile{
 	{"-p75", 75},
 	{"-p50", 50},
 }
+
+// vector holds the base vector implementation. This is meant to be embedded
+// by metric types that require a variable set of labels. Implements
+// PrometheusVector.
+type vector struct {
+	*syncutil.RWMutex
+	encounteredLabelsLookup map[string]struct{}
+	encounteredLabelValues  [][]string
+	orderedLabelNames       []string
+}
+
+func newVector(labelNames []string) vector {
+	sort.Strings(labelNames)
+
+	return vector{
+		RWMutex:                 &syncutil.RWMutex{},
+		encounteredLabelsLookup: make(map[string]struct{}),
+		encounteredLabelValues:  [][]string{},
+		orderedLabelNames:       labelNames,
+	}
+}
+
+func (v *vector) getOrderedValues(labels map[string]string) []string {
+	labelValues := make([]string, 0, len(labels))
+	for _, labelName := range v.orderedLabelNames {
+		labelValues = append(labelValues, labels[labelName])
+	}
+
+	return labelValues
+}
+
+// recordLabels records the given combination of label values if they haven't
+// been seen before. This is used to iterate over all the counters created
+// based on unique label combinations.
+func (v *vector) recordLabels(labelValues []string) {
+	v.RLock()
+	lookupKey := strings.Join(labelValues, "_")
+	if _, ok := v.encounteredLabelsLookup[lookupKey]; ok {
```

Comment:
i think the label key needs to be apart of the keys in `encounteredLabelsLookup`.

Suppose the user declares a counter vector annual salary with labels `HighSchoolDiploma` and `IsMale` and the user passes the label values true and false to both. If the user wants to update the salary metric with `HighSchoolDiploma==true` without specifying the `IsMale` metric value, there is no way to distinguish that from if the user just update metric with the `IsMale==true` label key value pair.

Reply:
I think this is working fine: we always put the value -- which could be empty string! -- for every label into this string. So one of those would be true_ an the other _true, but those are distinct strings that work just fine as map keys.

Reply:
> If the user wants to update the salary metric with HighSchoolDiploma==true without specifying the IsMale metric value, there is no way to distinguish that from if the user just updated the metric with the IsMale==true label key value pair.

When I started working on this, I was taking inspiration from the [MetricVec](https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#MetricVec) type in the prometheus go client. There it doesn't allow the above scenario (instrumenting with a partial set of labels). Which, I think, should be fine? 

Like @dt mentioned, it still works by using empty strings. We can maybe come up with a more permanent solution if there is a usecase for partial labels.